### ISSUE TITLE
CVE-2023-24815 - Directory Traversal vulnerability in io.vertx:vertx-web

### ISSUE THREAD

**Issue by abstractj** [keycloak/keycloak#17476]
### Description

**Note**: The issue affects the Keycloak Operator and the Quarkus distribution

-   Package Manager: maven
-   Vulnerable module: io.vertx:vertx-web
-   Introduced through: org.keycloak:[keycloak-operator@999.0.0-SNAPSHOT](mailto:keycloak-operator@999.0.0-SNAPSHOT), io.quarkiverse.operatorsdk:quarkus-operator-sdk@4.0.7 and others

### Detailed paths

-   _Introduced through_: org.keycloak:[keycloak-operator@999.0.0-SNAPSHOT](mailto:keycloak-operator@999.0.0-SNAPSHOT) › io.quarkiverse.operatorsdk:quarkus-operator-sdk@4.0.7 › io.quarkus:[quarkus-smallrye-health@2.13.7.Final](mailto:quarkus-smallrye-health@2.13.7.Final) › io.quarkus:[quarkus-vertx-http@2.13.7.Final](mailto:quarkus-vertx-http@2.13.7.Final) › io.vertx:vertx-web@4.3.4

# Overview

[io.vertx:vertx-web](https://github.com/vert-x3/vertx-web) is a HTTP web applications for Vert.x.

Affected versions of this package are vulnerable to Directory Traversal when running Vertx web applications that serve files using `StaticHandle`. If the mount point is a wildcard (`*`) then an attacker can exfiltrate any class path resource.

When computing the relative path to locate the resource, in case of wildcards, [the code](https://github.com/vert-x3/vertx-web/blob/62c0d66fa1c179ae6a4d57344631679a2b97e60f/vertx-web/src/main/java/io/vertx/ext/web/impl/Utils.java#L83) returns the user input (without validation) as the segment to lookup. Even though checks are performed to avoid escaping the sandbox, given that the input was not sanitized, `\` are not properly handled and an attacker can build a path that is valid within the classpath.

**Note:** This vulnerability can only be exploited on Windows Operating Systems and Windows File Systems

# Details

A Directory Traversal attack (also known as path traversal) aims to access files and directories that are stored outside the intended folder. By manipulating files with "dot-dot-slash (../)" sequences and its variations, or by using absolute file paths, it may be possible to access arbitrary files and directories stored on file system, including application source code, configuration, and other critical system files.

Directory Traversal vulnerabilities can be generally divided into two types:

-   **Information Disclosure**: Allows the attacker to gain information about the folder structure or read the contents of sensitive files on the system.

`st` is a module for serving static files on web pages, and contains a [vulnerability of this type](https://snyk.io/vuln/npm:st:20140206). In our example, we will serve files from the `public` route.

If an attacker requests the following URL from our server, it will in turn leak the sensitive private key of the root user.

```
curl http://localhost:8080/public/%2e%2e/%2e%2e/%2e%2e/%2e%2e/%2e%2e/root/.ssh/id_rsa
```

**Note** `%2e` is the URL encoded version of `.` (dot).

-   **Writing arbitrary files**: Allows the attacker to create or replace existing files. This type of vulnerability is also known as `Zip-Slip`.

One way to achieve this is by using a malicious `zip` archive that holds path traversal filenames. When each filename in the zip archive gets concatenated to the target extraction folder, without validation, the final path ends up outside of the target folder. If an executable or a configuration file is overwritten with a file containing malicious code, the problem can turn into an arbitrary code execution issue quite easily.

The following is an example of a `zip` archive with one benign file and one malicious file. Extracting the malicious file will result in traversing out of the target folder, ending up in `/root/.ssh/` overwriting the `authorized_keys` file:

```
2018-04-15 22:04:29 .....           19           19  good.txt
2018-04-15 22:04:42 .....           20           20  ../../../../../../root/.ssh/authorized_keys
```

# Remediation

Upgrade `io.vertx:vertx-web` to version 4.3.8 or higher.

# References

-   [GitHub Commit](https://github.com/vert-x3/vertx-web/commit/8a6b4ef8096a0b8714c1f10ec0133f4b42e48428)
-   [PoC](https://github.com/adrien-aubert-drovio/vertx-statichandler-windows-traversal-path-vulnerability)
-   [Vulnerable Code](https://github.com/vert-x3/vertx-web/blob/62c0d66fa1c179ae6a4d57344631679a2b97e60f/vertx-web/src/main/java/io/vertx/ext/web/impl/Utils.java#L83)

**Comment by abstractj** [comment]
Keycloak does not make any use of `StaticHandle` from vertx-web, thus we are not vulnerable, but affected.

**Comment by stianst** [comment]
release/21.1 is currently on 4.3.4 so this one is probably not resolved

**Comment by abstractj** [comment]
@stianst you are correct, and we can reopen the issue. According to the analysis from the Quarkus team: 
"Fix will be available with updating Quarkus to Vert.x 4.4.1 however it will not be possible to backport it to the 2.13.x product branch. 
Also, according to the team’s analysis, Quarkus is not relying on the vulnerable code"

**Comment by stianst** [comment]
@abstractj in that case I'm dropping it for the patch release and moving to 22, as we won't be able to resolve this in 21.1.x unless resolved in Quarkus 2.13, which doesn't sound like it's critical since it's not relying on the vulnerable code.
